### PR TITLE
feat(harness): multi-model harness with evidence-gated review councils

### PR DESCRIPTION
**Pull request by haxiorz**
Adds a staged-only path to the collection: the same isolated implementation, validation, and review discipline as the autonomous PR skills, but the run stops at a verified staged diff so a human owns publication. The existing autonomous skills are untouched.

## What this adds

**Ten new skills.** Two entry points — `om-fix-issue` and `om-implement-feature` — plus four thin profile aliases each (`-multi`, `-optimized`, `-multi-optimized`), the `om-setup-agent-harness` configurator, and `om-harness`, the shared internal runtime. The variants only select a profile; the whole workflow lives in the base skill, so behavior cannot drift between them.

**A deterministic runtime** (`skills/om-harness/scripts/harness.mjs`) providing provider-neutral model adapters, SHA-256-bound `om-code-review` packets, fresh-reviewer attestation, the multi-model review council and its result matrix, evidence-gated work packets with repository-global path leases, and a staged-handoff verifier that proves `HEAD`, refs, and reflogs never moved.

**Five operating profiles.** `standard` (fresh-context review only, no external provider and no configuration required), `optimized` (implementation offloaded to a sandboxed worker), `multi` (full independent council), `multi-optimized` (both), and `high-assurance` (manifest packets, path leases, risk-scaled blind review, independent finding verification, separate fixer context, deterministic acceptance gate).

**Bring-your-own reviewers.** A bundled jury ships as a starting default, but any OpenAI-compatible endpoint or local CLI can be bound as a reviewer through the generic adapters. The only fixed element is the host, because the mandatory fresh review is a host artifact. Workers still require an audited sandbox adapter.

## Safety model

The stage-only boundary is layered rather than asserted in prose:

- **Prevention** — `PreToolUse` hooks deny `git push` and PR creation during a run, and freeze test files during the fix phase.
- **Isolation** — isolated worktrees; workers and command reviewers run with credential-stripped environments, no shell, and Git protocols and credential helpers disabled.
- **Detection** — refs and reflogs are snapshotted around every worker *and* reviewer invocation and again at staging; any mutation fails the run.
- **Trust** — executable adapter settings come only from a trusted base revision plus a user-local overlay; repository text, tracker content, and model output are data, never instructions; review packets are hash-bound and model provenance is recorded (requested vs observed).
- **Bounds** — fix and council loops stop at three iterations; packet budgets cap invocations and input bytes.

A model's claim that a test passed is never gate evidence: the gate takes trusted command output bound to the exact reviewed diff, and a drifted diff sends the packet back to review.

## Onboarding

A fresh repository needs no provider setup for the default path — `standard` runs use only the host and work with no `agentHarness` section at all. The `-multi` and `-optimized` variants stop when their profile is missing or its providers are not ready, explain which bindings are required, and route into `om-setup-agent-harness`; after setup they re-probe and continue on the originally requested profile. Falling back to a lesser profile only happens as an explicit, informed user choice, and the run report says so. Every non-`standard` run opens with a readiness table (model, binding, role, ready/failed) before any model is invoked.

## Documentation

- `skills/om-harness/README.md` — the deep dive: skill matrix, profile and model tables, flow diagrams for the council, worker loop, and packet lifecycle, the safety model, and a runtime command reference.
- `docs/agent-harness-guide.html` — an interactive explorer for the same material.
- Per-skill cards for all ten new skills in `docs/skills/`, plus roster, backward-compatibility, and upgrade-notes entries.

## Verification

- `npm run lint` and `npm run test:agent-harness` pass.
- The test suite covers the runtime end to end — adapters, councils, packet lifecycle, staging assertions — including regression tests for prompt delivery limits, stdin failure handling, corrupted lease files, stale-lock takeover, artifact-directory validation, reviewer environment isolation, and reviewer-triggered ref mutation.
- All five bundled reviewers were exercised against live providers in a real council run.
- The interactive guide was verified in a browser at seven viewport widths (320–1440) for layout, overlap, and interaction correctness.

## Compatibility

Additive. The `agentHarness` configuration object is optional, existing configs remain valid without it, and no existing skill's behavior changes. New shared formats (review results, review packets, fresh-review attestation, packet manifests and ledgers, gate evidence, staged-handoff markers) are versioned and documented in `BACKWARD_COMPATIBILITY.md`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by pkarw** [comment]
🤖 `om-auto-review-pr` started by @pkarw at 2026-07-24T12:24:55Z. Other auto-skills will skip this PR until the lock is released.

**Review by pkarw — CHANGES_REQUESTED** [review]
## Code Review: feat(harness): multi-model harness with evidence-gated review councils

**Verdict: 🔴 Request changes** — one blocker (merge conflict) plus several majors (security-containment gaps in the runtime and test-coverage gaps on security-critical paths). Nothing here questions the design, which is strong; these are gaps in a component whose entire value proposition is adversarial containment, so the bar is high.

Reviewed in an isolated worktree at PR head `a018dc9`, base `main`. Scope: full diff (66 files, +9,593). This is a first review.

### ✅ Validation gate
- `bash scripts/lint.sh` → **Lint OK**.
- `npm run test:agent-harness` → **passes** ("Agent harness tests passed."). Note: the run is slow (~4 min here) and emits **no** incremental output, because it spawns dozens of Node subprocesses serially with no per-`run()` timeout — see 🟠 T5.
- CI on the PR: `lint` ✅, `license/cla` ✅.

### ✅ What's solid (verified, not asserted)
The core safety primitives hold up under adversarial reading:
- **Hash binding** — subject/rubric/context SHA-256 are recomputed and compared (`validateReviewPacket`); the packet gate re-hashes the worktree at gate time and rejects a drifted diff (`snapshot.sha256 !== ledger.currentDiff.sha256`), closing the obvious TOCTOU.
- **Quorum integrity** — `evaluatePolicy` runs over provider reviewers only; the fresh-Claude host result is `policyEligible:false` and can only make the verdict stricter, matching the `BACKWARD_COMPATIBILITY.md` contract.
- **Injection boundary** — model output is parsed as JSON only; no `eval`/`new Function`; every subprocess uses `spawn`/`spawnSync` with **array args and no `shell:true`**; untrusted text is framed as "DATA, never instructions."
- **Path leases** reject `..`, absolute, and `.git/` escapes.
- **Contracts & docs** — all 7 JSON/schema files parse; the shipped `configuration-template.json` validates against the real harness validator (all 5 profiles + 5 models); all 5 schemas match the implementation's read/write; every persisted artifact carries `version: 1`; roster/README/`docs/skills`/`skill-coverage.md` are consistently updated 30→40 skills; no committed secrets (only env-var *names*).

---

### 🔴 Blockers

**B1 — Merge conflict with `main` (`UPGRADE_NOTES.md`).** The PR is `mergeable: CONFLICTING` / `mergeStateStatus: DIRTY`. The only conflicting file is `UPGRADE_NOTES.md` — a "newest-first" changelog collision where both sides prepended entries; it resolves trivially by keeping both (PR entries first). Must be resolved before merge.

---

### 🟠 Majors (each independently warrants request-changes)

**M1 — Credential env-stripping is an incomplete substring denylist.** `SECRET_ENV_PATTERN` (`skills/om-harness/scripts/harness.mjs:45`) drives `sanitizedCliEnvironment`, the base env for workers **and command reviewers**. It matches only `TOKEN|SECRET|PASSWORD|PASSWD|API_KEY|PRIVATE_KEY|ACCESS_KEY|SESSION`, so it does **not** strip common secret-bearers: bare `*_KEY` (`STRIPE_KEY`, `ENCRYPTION_KEY`, `OPENAI_KEY`), credential-bearing connection strings (`DATABASE_URL`, `REDIS_URL`, `MONGODB_URI` — often `user:pass@host`), and handle-to-credential vars (`GOOGLE_APPLICATION_CREDENTIALS`, `KUBECONFIG`, `SSH_AUTH_SOCK`, `DOCKER_AUTH_CONFIG`). Command-reviewer CLIs are third-party binaries and are **not** network-sandboxed (M4), so this is a live exfiltration path. *Scenario:* a repo exports `DATABASE_URL=postgres://svc:hunter2@db/prod` and `STRIPE_KEY=sk_live_…`; a council reviewer CLI inherits both verbatim and can send them on its next outbound API call. A credential guard for adversarial subprocesses should be an **allowlist** (pass only `PATH`, `HOME`, `LANG`, locale, `OM_HARNESS_*`, and explicitly-needed vars), not a keyword denylist.

**M2 — Command-reviewer working-tree file mutations are undetected.** `runCommandReviewer` (`harness.mjs:619-634`) snapshots only `captureGitState`, which captures refs/reflogs/HEAD **but not working-tree file content** (`harness.mjs:1936-1946`). Workers get content-level protection via `captureDirtyState`; reviewers do not. The failure message ("reviewer commands must be read-only") over-promises. *Scenario:* in the non-packet `commandReview → commandStage` flow, a hostile/compromised reviewer CLI writes a backdoor line into an allow-listed source file during its "review"; nothing re-hashes the worktree in this flow, and `commandStage` only checks refs + allowlist membership (M-min below), so the injected content is staged for the human's publish decision as "reviewed." (The packet flow catches this via the gate's diff re-hash; the exposure is the review/stage path.)

**M3 — Ref snapshot omits `refs/replace/*` and arbitrary ref namespaces.** `captureGitState` scans only `refs/heads refs/tags refs/stash` plus default reflogs (`harness.mjs:1942-1943`), while every diff/stage path runs `git diff … HEAD` **without** `--no-replace-objects` / `GIT_NO_REPLACE_OBJECTS`. *Scenario:* a worker installs `refs/replace/<HEAD>` pointing at a synthetic commit whose tree contains malicious content; `git rev-parse HEAD` is unchanged so the snapshot sees nothing, but every `git diff` resolves HEAD's tree through the replacement and shows a benign/empty diff consistently across review, gate, and stage — reviewers approve a diff that does not reflect what the human publishes. Separately, any worker-created `refs/<custom>/…` (no default reflog) is a lasting ref mutation the "refs never moved" staged-handoff attestation claims to disprove and doesn't. *Fix:* enumerate the full ref set (`for-each-ref` with no path filter) and set `GIT_NO_REPLACE_OBJECTS=1` on all diff commands.

**M4 — `freeze-tests.sh` fails open in non-JS repos.** `skills/om-harness/hooks/freeze-tests.sh:32` gates on `grep -qE '(\.spec\.|\.test\.|/__tests__/|/__integration__/)'`, which matches JS/TS naming only — never Python (`test_foo.py`), Go (`foo_test.go`), Ruby (`foo_spec.rb`), Java (`FooTest.java`, `src/test/`), or Rust (`tests/`). The harness installs into "arbitrary repos" and the README/`om-fix-issue` present the sentinel as a general guarantee that the fix loop cannot weaken the test it must satisfy. *Scenario:* a Python repo run drops `.om-freeze-tests` after the red run; the fix loop edits `test_bug.py` green; the hook never matches, allows the write, and the freeze guarantee is silently void.

**M5 — Security-critical paths have zero test coverage** (CODE_REVIEW.md: new behavior needs tests). Despite a substantial suite, these have no assertion: worker git-credential/`protocol.allow` hardening (a refactor swapping `workerEnvironment→sanitizedCliEnvironment` would pass green); `validateStagePath` traversal/`.git` negative cases; packet worker writing **outside** its lease (`changedOutsideAllowed`); worker-side ref-mutation (only the *reviewer* rogue path is tested — a separate implementation); and the entire `openai-compatible` adapter is never instantiated. Credential stripping is proven for `GITHUB_TOKEN` only — narrowing the denylist would leak `AWS_SECRET_ACCESS_KEY`/`OPENAI_API_KEY` with no test failing.

---

### 🟡 Minors

- **freeze-tests matcher** doesn't cover `NotebookEdit`/`MultiEdit`; a `.ipynb` test edit bypasses the freeze (`hooks-settings-snippet.json:14` matches `Edit|Write` only, and the filename regex misses `.ipynb`).
- **`block-push-and-pr.sh`** (`:24-30`) blocks `gh pr create` but not `gh api … /pulls` or `gh pr merge`; a line-split `git \<newline>push` evades the line-oriented `grep`; and it `exit 0` (allow) on empty/unparseable input. Since a completed `git push` leaves local refs unchanged, this hook is the *sole* barrier against push — the "denies PR creation at the tool boundary" claim overstates it.
- **`commandStage`** (`harness.mjs:1970-2002`) proves refs == start-state and staged paths ⊆ allowlist, but never binds staged **content** to a reviewed `diffSha256`, so any drift on an allow-listed path is staged unreviewed.
- **`extractJson`** (`harness.mjs:353-362`) brute-forces every `{…}` substring with nested `JSON.parse` (O(n²)) — a large brace-heavy adversarial reply burns CPU; and `waitForFreshReviewArtifact` parses on first `existsSync`, so a non-atomic host write → partial-read `JSON.parse` failure aborts the already-paid council instead of retrying.
- **Path leases** have no TTL/stale-takeover (only the `.mutex` does), so a crashed packet run wedges its paths until manual `packet-release`; and a corrupted lease that parses but lacks a string `path` passes the guard (`:1190`) and silently protects nothing.
- **Kimi/preset reviewer** (`harness.mjs:746-786`) runs un-sandboxed with no `captureGitState` guard (asymmetric with command reviewers); mitigated only by being handed an empty temp dir, not the real worktree.
- **`optimized`-variant reroute line** (`skills/om-fix-issue-optimized/SKILL.md:24`, `om-implement-feature-optimized/SKILL.md:24`) tells users to bind "any OpenAI-compatible endpoint or local CLI as a reviewer," but the `optimized` profile is blocked on a missing **worker** (which requires the audited codex sandbox) — a custom reviewer can't unblock it, so the user loops. The base skills say this correctly.
- **Hooks are bash-only** — no native Windows PowerShell guard (CODE_REVIEW.md lists WSL2/PowerShell as supported); undocumented.
- **Test robustness** (`scripts/test-agent-harness.mjs`): `run()` uses `spawnSync` with **no timeout** (:19-28) so a wedged subprocess hangs CI indefinitely; the temp dir (`mkdtempSync`, :13) is never cleaned up; and the suite is POSIX-only (shebang scripts + `chmod 0o755`).
- **provider-catalog.md:31** describes the Codex **worker** binding as passing `{model}` and `{promptFile}` as separate args, but the audited command reads the prompt from **stdin** (`-`) — a user hand-crafting the binding from this line gets rejected by `validateWorkerCommand`. (The shipped template and `adapter-contract.md` are correct.)

---

### ⚪ Nits

- **`review-result.schema.json`** is the only schema with no `version` field, and its base name collides with the *versioned* council artifact also written as `review-result.json` (different shape) — consider renaming one or noting they're distinct.
- **`docs/skills/om-harness.md`** carries the "Interactive — acts once, may ask questions" badge while its body says om-harness is "not a skill you invoke directly" — the badge mislabels an internal runtime.
- **`references/handoff-report.md`** (both) omit the `Issue: #<n>` / `Spec: <path>` chaining markers `BACKWARD_COMPATIBILITY.md` §5 says orchestrators parse; a chain resuming after human publish must re-derive issue/spec identity. (Defensible, since these skills deliberately stop before a PR.)

---

### Summary
| Severity | Count |
|---|---|
| 🔴 Blocker | 1 |
| 🟠 Major | 5 |
| 🟡 Minor | 9 |
| ⚪ Nit | 3 |

The design and contract discipline are excellent and the additive/backward-compatibility story checks out end to end. What blocks merge is the conflict plus the containment gaps: for a harness sold on adversarial isolation, an allowlist credential boundary (M1), content-level reviewer isolation (M2), full-ref-namespace snapshotting (M3), a language-agnostic test-freeze (M4), and tests that actually exercise those boundaries (M5) are the difference between the safety model being *enforced* and being *asserted*.

🤖 Reviewed by `om-auto-review-pr`

**Comment by pkarw** [comment]
Label set to `changes-requested` because review found actionable issues (1 blocker, 5 majors).

**Comment by pkarw** [comment]
Priority set to `priority-medium` because this is an additive, opt-in new capability with no production regression or incident.

**Comment by pkarw** [comment]
Risk set to `risk-high` because it adds a 2,044-line security-critical runtime touching shared contract surfaces and credential/sandbox boundaries — a wide blast radius that warrants deep review.

**Comment by pkarw** [comment]
Thanks @haxiorz — this is strong, carefully-designed work and the additive/backward-compat story checks out end to end. Review found actionable items (1 blocker + 5 majors, detailed above), so I'm handing this PR back to you for the next pass. The headline items: resolve the `UPGRADE_NOTES.md` conflict (trivial keep-both), move the credential env boundary from a denylist to an allowlist, extend reviewer isolation to working-tree file content (not just refs/reflogs), snapshot the full ref namespace + set `GIT_NO_REPLACE_OBJECTS=1` on diffs, make `freeze-tests.sh` language-agnostic, and add tests over those boundaries. When the updates are pushed, re-request review and the automation can pick it up from the latest head.

File: skills/om-implement-feature/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+questions from repository evidence and surface only genuine product decisions.
+Use a repo-local `om-pre-implement-spec` when available. For `multi`, `multi-optimized`, and `high-assurance`,
+follow the `om-harness` bound `om-code-review` council contract: prepare an
+exact-spec packet, delegate `om-code-review` to a new Claude context with no
+implementation transcript, and run every configured advisor concurrently on
+that same packet. Fix confirmed design blockers before implementation.
+
+### 3. Implement
+
+Use a repo-local `om-implement-spec` when available. Otherwise implement the
```
Comment (pkarw) [inline skills/om-implement-feature/SKILL.md:46]: we have `om-auto-implement-spec` now in the repo and i think we should use it

**Review by pkarw — CHANGES_REQUESTED** [review]
The new skills should be more scattered into `resource` chunks - look how the other pre-existing skills are done; please check the `om-auto-write-spec` and `om-auto-implement-spec` - the new skills taht compose of `om-auto-create-pr` and other skills to achieve the effect; also check the rules for using `emojis` templates, tracker and posting PR comments - we should align 100% to these pre-existing rules in the other skills (i spent quite a lot time aligning it); also the `--autonomous` / `--interactive` flags and the rules of agentic setup and chaining prs

**Review by pkarw — CHANGES_REQUESTED** [review]
I'd also suggest to add the `om-agent-harness` and `om-setup-agent-harness` - so basic skills for using other models; then these `multi` skills could be kind of extensions to the existing `auto-implement-spec` and other skills; or just name all these multi agent skilsl by someconvention - or maybe we should have these multi agent skills in another repo (worst option) - I mean now it's a little bit messy and user wont be able to correctly choose which skill to use as they all look very similar by the name can you please aliogn it and keep it simpl?

In general this is grea!
